pkg/steps/template_test.go

```diff
@@ -0,0 +1,173 @@
+package steps
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/openshift/ci-operator/pkg/api"
+	"k8s.io/apimachinery/pkg/util/diff"
+)
+
+func TestDeferredParametersAllLinks(t *testing.T) {
+	var testCases = []struct {
+		purpose       string
+		dp            *DeferredParameters
+		expectedItems int
+	}{{
+		purpose: "sanity",
+		dp: &DeferredParameters{
+			links: map[string][]api.StepLink{
+				"K1": {someStepLink("ONE"), someStepLink("TWO")},
+				"K2": {someStepLink("THREE")},
+			},
+		},
+		// only compare count of returned items because AllLinks can
+		// return links in any order
+		expectedItems: 3,
+	}}
+
+	for _, tc := range testCases {
+		links := tc.dp.AllLinks()
+
+		if len(links) != tc.expectedItems {
+			t.Errorf("%s: %v.AllLinks() returned %v, expected %d items", tc.purpose, tc.dp, links, tc.expectedItems)
+		}
+	}
+}
+
+func TestDeferredParametersMap(t *testing.T) {
```

Comment:
Not being super familiar with the underlying code, I'm having trouble understanding the test. If we had more than one case it would be useful here, also a description/name that provided setup --> end result causality.

Is this test case ensuring that a func doesn't override a previous value? Does it enforce that a previous value without a func is not exposed? Are there other cases? Can we break them up into separate cases?

Reply:
I'll try to look into this to make it more readable, at least with descriptions and/or names.

Reply:
I've broken the testcase to smaller, more targeted ones. You were right I was probably trying to do too much in a single one.